# Week 2 Day 05 — Transformers

## Objectives

- Use Hugging Face pipeline() for sentiment analysis
- Use pipeline() for Named Entity Recognition (NER)
- Use pipeline() for text summarization
- Use pipeline() for zero-shot classification
- Inspect tokenizer output on tricky text
- Compare two pretrained models on the same input

In [1]:
from transformers import pipeline

In [2]:
sentiment_pipeline = pipeline("sentiment-analysis")

[transformers] No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f.
Using a pipeline without specifying a model name and revision in production is not recommended.


Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

In [3]:
text = "I really enjoyed learning about Transformers today."

result = sentiment_pipeline(text)

print(result)

[{'label': 'POSITIVE', 'score': 0.9996664524078369}]


In [4]:
texts = [
    "I love this course.",
    "This is an amazing Transformer model.",
    "I hate this product.",
    "This was a terrible experience.",
    "The book is on the table."
]

results = sentiment_pipeline(texts)

for text, result in zip(texts, results):
    print(text)
    print(result)
    print()

I love this course.
{'label': 'POSITIVE', 'score': 0.9998787641525269}

This is an amazing Transformer model.
{'label': 'POSITIVE', 'score': 0.9998767375946045}

I hate this product.
{'label': 'NEGATIVE', 'score': 0.9997596144676208}

This was a terrible experience.
{'label': 'NEGATIVE', 'score': 0.9989833235740662}

The book is on the table.
{'label': 'POSITIVE', 'score': 0.9883579611778259}



In [1]:
print("Jupyter kernel is working")

Jupyter kernel is working


In [2]:
from transformers import pipeline

print("Transformers imported successfully")

Transformers imported successfully


In [3]:
ner_pipeline = pipeline(
    "ner",
    model="dslim/distilbert-NER",
    aggregation_strategy="simple"
)

print("NER pipeline loaded successfully")

config.json:   0%|          | 0.00/926 [00:00<?, ?B/s]

D:\arun\ai-learning\.venv\Lib\site-packages\huggingface_hub\file_download.py:141: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\god\.cache\huggingface\hub\models--dslim--distilbert-NER. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


model.safetensors: reconstructing file:   0%|          |  0.00B /  261MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/102 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/1.20k [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/669k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

NER pipeline loaded successfully


In [4]:
text = """
Rajesh works at Microsoft in Chennai.
He is learning Artificial Intelligence and Natural Language Processing.
"""

entities = ner_pipeline(text)

for entity in entities:
    print(entity)

{'entity_group': 'PER', 'score': np.float32(0.9984669), 'word': 'Raj', 'start': 1, 'end': 4}
{'entity_group': 'PER', 'score': np.float32(0.9979146), 'word': '##esh', 'start': 4, 'end': 7}
{'entity_group': 'ORG', 'score': np.float32(0.9976775), 'word': 'Microsoft', 'start': 17, 'end': 26}
{'entity_group': 'LOC', 'score': np.float32(0.99336195), 'word': 'Chennai', 'start': 30, 'end': 37}
{'entity_group': 'ORG', 'score': np.float32(0.5953299), 'word': 'Art', 'start': 54, 'end': 57}
{'entity_group': 'ORG', 'score': np.float32(0.63866204), 'word': '##ific', 'start': 57, 'end': 61}
{'entity_group': 'ORG', 'score': np.float32(0.54869366), 'word': '##ial Intelligence', 'start': 61, 'end': 77}
{'entity_group': 'MISC', 'score': np.float32(0.7025595), 'word': 'Natural Language Processing', 'start': 82, 'end': 109}


### NER Observation

The model attempts to identify named entities in the sentence.

For example:

- Rajesh → Person
- Microsoft → Organization
- Chennai → Location

The exact labels, confidence scores, and recognized entities depend on
the pretrained model.

## 4. Text Summarization

Text summarization creates a shorter version of a longer piece of text
while attempting to preserve the important information.

We will use a pretrained Transformer summarization model.

In [9]:
from transformers import pipeline

text_generator = pipeline(
    "text-generation",
    model="distilgpt2"
)

print("Text generation pipeline loaded successfully")

config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

D:\arun\ai-learning\.venv\Lib\site-packages\huggingface_hub\file_download.py:141: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\god\.cache\huggingface\hub\models--distilgpt2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


model.safetensors: reconstructing file:   0%|          |  0.00B /  353MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/76 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Text generation pipeline loaded successfully


In [10]:
article = """
Artificial intelligence is changing many industries around the world.
Machine learning allows computers to learn patterns from data without
being explicitly programmed for every individual task. Natural language
processing focuses on enabling computers to understand and work with
human language. Modern Transformer models have significantly improved
many NLP applications including translation, summarization, question
answering, and text generation.
"""

prompt = f"""
Summarize this article briefly:

{article}

Summary:
"""

result = text_generator(
    prompt,
    max_new_tokens=60,
    do_sample=False
)

print("Generated text:")
print(result[0]["generated_text"])

[transformers] Passing `generation_config` together with generation-related arguments=({'do_sample', 'max_new_tokens'}) is deprecated and will be removed in future versions. Please pass either a `generation_config` object OR all generation parameters explicitly, but not both.
[transformers] Both `max_new_tokens` (=60) and `max_length`(=50) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
[transformers] Ignoring clean_up_tokenization_spaces=True for BPE tokenizer GPT2Tokenizer. The clean_up_tokenization post-processing step is designed for WordPiece tokenizers and is destructive for BPE (it strips spaces before punctuation). Set clean_up_tokenization_spaces=False to suppress this warning, or set clean_up_tokenization_spaces_for_bpe_even_though_it_will_corrupt_output=True to force cleanup anyway.


Generated text:

Summarize this article briefly:


Artificial intelligence is changing many industries around the world.
Machine learning allows computers to learn patterns from data without
being explicitly programmed for every individual task. Natural language
processing focuses on enabling computers to understand and work with
human language. Modern Transformer models have significantly improved
many NLP applications including translation, summarization, question
answering, and text generation.


Summary:
The most important thing about artificial intelligence is that it is not just a tool for learning, but a tool for learning.
The most important thing about artificial intelligence is that it is not just a tool for learning, but a tool for learning.
The most important thing about artificial intelligence is that it


### Summarization Observation

The text-generation model generates text based on the provided prompt.

However, DistilGPT2 is a general text-generation model rather than a
model specifically fine-tuned for summarization.

This demonstrates that choosing an appropriate pretrained model for a
specific NLP task is important.

## 5. Zero-Shot Classification

Zero-shot classification allows us to classify text into categories
provided at inference time.

The model can compare the input text with candidate labels without
requiring us to train a new classifier for those labels.

In [1]:
from transformers import pipeline

classifier = pipeline(
    "zero-shot-classification",
    model="valhalla/distilbart-mnli-12-1"
)

print("Zero-shot classification pipeline loaded successfully")

D:\arun\ai-learning\.venv\Lib\site-packages\huggingface_hub\file_download.py:141: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\god\.cache\huggingface\hub\models--valhalla--distilbart-mnli-12-1. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


Zero-shot classification pipeline loaded successfully


In [2]:
text = "I am learning Python, machine learning, and neural networks."

candidate_labels = [
    "technology",
    "sports",
    "politics",
    "cooking"
]

result = classifier(
    text,
    candidate_labels
)

print("Text:")
print(text)

print("\nLabels:")
print(result["labels"])

print("\nScores:")
print(result["scores"])

Text:
I am learning Python, machine learning, and neural networks.

Labels:
['technology', 'cooking', 'sports', 'politics']

Scores:
[0.9946146011352539, 0.002004363341256976, 0.001856629503890872, 0.0015243695816025138]


### Zero-Shot Observation

The model receives:

1. The input text
2. A list of candidate categories

It then ranks the categories according to how well they match the text.

For this example, the `technology` category should generally receive a
high score because the text discusses Python, machine learning, and
neural networks.

In [3]:
examples = [
    "The football team won the championship.",
    "Python is widely used in machine learning.",
    "The government announced a new policy."
]

candidate_labels = [
    "sports",
    "technology",
    "politics"
]

for text in examples:
    result = classifier(
        text,
        candidate_labels
    )

    print("Text:", text)
    print("Prediction:", result["labels"][0])
    print("Score:", result["scores"][0])
    print()

Text: The football team won the championship.
Prediction: sports
Score: 0.9914373755455017

Text: Python is widely used in machine learning.
Prediction: technology
Score: 0.9933564066886902

Text: The government announced a new policy.
Prediction: politics
Score: 0.8929926753044128



## 6. Tokenization

Before text is processed by a Transformer, raw text must be converted
into tokens.

A tokenizer converts text into tokens and then maps those tokens to
numerical token IDs that the model can process.

In [5]:
from transformers import AutoTokenizer

print("AutoTokenizer imported successfully")

AutoTokenizer imported successfully


In [6]:
tokenizer = AutoTokenizer.from_pretrained(
    "bert-base-uncased"
)

print("Tokenizer loaded successfully")

D:\arun\ai-learning\.venv\Lib\site-packages\huggingface_hub\file_download.py:141: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\god\.cache\huggingface\hub\models--bert-base-uncased. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


Tokenizer loaded successfully


In [7]:
text = "I love learning NLP!"

tokens = tokenizer.tokenize(text)

print("Original text:")
print(text)

print("\nTokens:")
print(tokens)

Original text:
I love learning NLP!

Tokens:
['i', 'love', 'learning', 'nl', '##p', '!']


### Understanding Subword Tokenization

A tokenizer does not always represent an entire word as one token.

Some words can be split into smaller subword pieces.

This allows Transformer models to handle uncommon words and words that
may not appear as complete vocabulary entries.

## 7. Inspect Tokenization on Tricky Text

We will test punctuation, contractions, hyphenated words, and uncommon
words to observe how the tokenizer handles them.

In [8]:
tricky_text = """
Transformers are amazing!
I can't believe NLP models understand words like unbelievable,
AI-powered, and Chennai-based.
"""

tokens = tokenizer.tokenize(tricky_text)

print("Original text:")
print(tricky_text)

print("\nTokens:")
print(tokens)

Original text:

Transformers are amazing!
I can't believe NLP models understand words like unbelievable,
AI-powered, and Chennai-based.


Tokens:
['transformers', 'are', 'amazing', '!', 'i', 'can', "'", 't', 'believe', 'nl', '##p', 'models', 'understand', 'words', 'like', 'unbelievable', ',', 'ai', '-', 'powered', ',', 'and', 'chennai', '-', 'based', '.']


In [9]:
encoded = tokenizer(tricky_text)

print("Encoded output:")
print(encoded)

Encoded output:
{'input_ids': [101, 19081, 2024, 6429, 999, 1045, 2064, 1005, 1056, 2903, 17953, 2361, 4275, 3305, 2616, 2066, 23653, 1010, 9932, 1011, 6113, 1010, 1998, 12249, 1011, 2241, 1012, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}


### Understanding the Encoded Output

The tokenizer output may contain:

- `input_ids` — numerical IDs representing tokens
- `attention_mask` — identifies valid input positions
- `token_type_ids` — identifies input segments for models that use them

The exact fields depend on the tokenizer and model.

In [10]:
print("Input IDs:")
print(encoded["input_ids"])

print("\nAttention Mask:")
print(encoded["attention_mask"])

Input IDs:
[101, 19081, 2024, 6429, 999, 1045, 2064, 1005, 1056, 2903, 17953, 2361, 4275, 3305, 2616, 2066, 23653, 1010, 9932, 1011, 6113, 1010, 1998, 12249, 1011, 2241, 1012, 102]

Attention Mask:
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


In [11]:
decoded_tokens = tokenizer.convert_ids_to_tokens(
    encoded["input_ids"]
)

print("Tokens reconstructed from IDs:")
print(decoded_tokens)

Tokens reconstructed from IDs:
['[CLS]', 'transformers', 'are', 'amazing', '!', 'i', 'can', "'", 't', 'believe', 'nl', '##p', 'models', 'understand', 'words', 'like', 'unbelievable', ',', 'ai', '-', 'powered', ',', 'and', 'chennai', '-', 'based', '.', '[SEP]']


In [12]:
decoded_text = tokenizer.decode(
    encoded["input_ids"]
)

print("Decoded text:")
print(decoded_text)

Decoded text:
[CLS] transformers are amazing! i can ' t believe nlp models understand words like unbelievable, ai - powered, and chennai - based. [SEP]


## 8. Compare Two Pretrained Models

Different pretrained models can produce different predictions for the
same input.

We will compare two sentiment-analysis models using exactly the same
sentence.

In [2]:
from transformers import pipeline

print("Pipeline imported successfully")

Pipeline imported successfully


In [3]:
sentiment_model_1 = pipeline(
    "sentiment-analysis",
    model="distilbert-base-uncased-finetuned-sst-2-english"
)

print("Model 1 loaded successfully")

Model 1 loaded successfully


In [4]:
sentiment_model_2 = pipeline(
    "text-classification",
    model="textattack/bert-base-uncased-SST-2"
)

print("Model 2 loaded successfully")

D:\arun\ai-learning\.venv\Lib\site-packages\huggingface_hub\file_download.py:141: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\god\.cache\huggingface\hub\models--textattack--bert-base-uncased-SST-2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


Model 2 loaded successfully


In [6]:
text = "The movie was surprisingly good, although the ending was disappointing."

result_1 = sentiment_model_1(text)
result_2 = sentiment_model_2(text)

print("Input:")
print(text)

print("\nModel 1:")
print(result_1)

print("\nModel 2:")
print(result_2)

Input:
The movie was surprisingly good, although the ending was disappointing.

Model 1:
[{'label': 'POSITIVE', 'score': 0.9996873140335083}]

Model 2:
[{'label': 'LABEL_1', 'score': 0.9602887630462646}]


## 8. Model Comparison Observation

Both pretrained models received exactly the same input sentence.

Model 1 predicted the sentiment as POSITIVE with a confidence score
of approximately 99.97%.

Model 2 produced LABEL_1 with a confidence score of approximately
96.03%.

The models produced different label names because different pretrained
models may use different label mappings. Therefore, label names such as
LABEL_0 and LABEL_1 should be checked using the model configuration
before interpreting them as positive or negative.

The confidence scores are also different because the models were trained
with different architectures, parameters, and training procedures.

This demonstrates that different pretrained Transformer models can
produce different outputs for the same input.

## 9. Overall Observations

### Sentiment Analysis

A pretrained Transformer can classify the sentiment of text without
training a new model from scratch.

### Named Entity Recognition

A pretrained NER model can identify entities such as people,
organizations, and locations.

### Text Generation and Summarization

A text-generation model can generate text from a prompt. However,
dedicated summarization models are generally more suitable for
high-quality summarization.

### Zero-Shot Classification

A pretrained model can classify text into categories supplied at
inference time.

### Tokenization

Raw text must be converted into tokens and numerical token IDs before
being processed by a Transformer.

### Model Comparison

Different pretrained models can behave differently on the same input.
Their predictions and confidence scores depend on their architecture,
training data, and learned parameters.

# Key Learnings

- Hugging Face `pipeline()` provides a simple interface for using
  pretrained Transformer models.
- Sentiment analysis predicts the sentiment expressed in text.
- NER identifies named entities such as people, organizations, and
  locations.
- Text-generation models generate text based on an input prompt.
- Dedicated models are generally preferred when a specific task such
  as summarization is required.
- Zero-shot classification allows text to be classified using
  user-provided categories.
- Tokenizers convert raw text into tokens and numerical token IDs.
- Subword tokenization allows models to represent uncommon words by
  splitting them into smaller pieces.
- `attention_mask` identifies valid input positions.
- Different pretrained models can produce different predictions and
  confidence scores for the same input.
- Pretrained Transformers allow us to perform useful NLP tasks without
  training models from scratch.